Implement pkg/tagset

djmitche opened the pull request:
This implements roughly the API in the included RFC.  A few details changed during the implementation period, but nothing critical.

### Motivation

See RFC.  This is the first portion of Phase 1.

### Additional Notes

As indicated in the commit message, this new code is completely unused elsewhere in the codebase.  The existing `HashedTags` code, which has a similar but not identical API, is moved to ./pkg/tagset/old.  The next step will be to modify the consumers of that API to use pkg/tagset instead.

### Describe how to test/QA your changes

Nothing to do, as new code is unused.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

djmitche committed:
Add new pkg/tagset types & functions

djmitche committed:
Move existing pkg/tagset to pkg/tagset/old

This will be the first thing to get converted to use the new pkg/tagset

This also copies SortUniqInPlace into the new package, in anticipation
of its being included in a module.  Since all of pkg/tagset/old will be
removed shortly, this duplication is not a problem.

mx-psi commented on go.mod:
```diff
@@ -64,6 +65,7 @@ require (
 	github.com/DataDog/datadog-agent/pkg/otlp/model v0.33.0-rc.4
 	github.com/DataDog/datadog-agent/pkg/quantile v0.33.0-rc.4
 	github.com/DataDog/datadog-agent/pkg/security/secl v0.33.0-rc.4
+	github.com/DataDog/datadog-agent/pkg/tagset v0.33.0-rc.4
```
We wanted to have this as a module so as to use it on the `pkg/otlp/model` module, to import that module on the OpenTelemetry Collector Datadog exporter. I personally don't want to import the `pkg/tagset` module directly from outside this repo. so we could make this module ['internal'](https://dave.cheney.net/2019/10/06/use-internal-packages-to-reduce-your-public-api-surface) for now. 

To be more explicit: if we make the path `pkg/internal/tagset` it would still fit the otlp use case (since it would be a separate module), but it would only be directly importable from modules/packages inside the datadog-agent repo (more concretely, modules/packages inside the `pkg` folder).

djmitche replied:
I think `pkg/internal` wouldn't work, as it'd be inaccessible from `cmd/`?

Anyway, since we don't use `internal` anywhere else in the codebase, I'd prefer not to introduce the idea with this project.  In particular, I think that use of `internal` is probably the right way to solve our concerns around other people linking to our modules in hopes of a stable API.  Which means probably almost all of `pkg` and `cmd` should move under `internal`, with the exception of things that we *do* support externally.

Also, I've designed pkg/tagset to have a stable API, so I would lean toward supporting its use externally.

In other words, I think `internal` is a bigger conversation.

mx-psi commented on pkg/aggregator/ckey/key_test.go:
```diff
@@ -9,7 +9,7 @@ import (
 	"testing"
 
-	"github.com/DataDog/datadog-agent/pkg/tagset"
+	tagset "github.com/DataDog/datadog-agent/pkg/tagset/old"
```
We use `tagset` in some places and `oldtagset` in others, what's the criteria?

djmitche replied:
No real criteria, although I'll make it consistent.  I suspect goimport did this?  Anyway, all of this `old` stuff is going away in the next PR.

mx-psi commented on pkg/tagset/base_factory.go:
```diff
@@ -0,0 +1,43 @@
+// implementations.
+type baseFactory struct {
+	// builders is a cache of unused builder instances for reuse
+	builders      []*Builder
```
Why not use a `sync.Pool` for this? If there is a good reason not to use it, I think we should mention it in a comment

djmitche replied:
`sync.Pool` is basically this, but with synchronization around it.  We don't need that synchronization since this is a single-thread data structure.

mx-psi replied:
`sync.Pool`'s implementation seems to be lock-free at least in the fast path (see e.g. golang/go@d5fd2dd6a17a816b7dfd99d4df70a85f1bf0de31). I definitely don't understand all the details but it's not obvious to me that this approach is clearly better. 

Anyway, not something to solve here, but maybe it warrants a comment?

djmitche replied:
atomics are also synchronization primitives, and have a cost in terms of processor synchronization.

But, I can add a comment.

djmitche replied:
Ah, I am thinking only of overhead, forgetting that `sync.Pool` also plays nicely with the gc and will drop all items when there is gc pressure.  I *think* there will only ever be one or two builders on this list, so not a great source of memory use, but we can modify it later if that assumption proves untrue.

mx-psi commented on pkg/tagset/default.go:
```diff
@@ -0,0 +1,55 @@
+// DefaultFactory is a global thread-safe factory, used by calls to
+// package-level functions.  This is suitable for non-performance-critical tags
+// manipulation
+var DefaultFactory Factory
```
Do we need to export the `DefaultFactory` or would it be enough to just export the functions below?

djmitche replied:
I suspect there are some cases where it's useful to use the DefaultFactory.  This is similar to `net/http`'s `DefaultServeMux`, for example.

mx-psi commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,176 @@
+// searching for cached instances.  A value of 5 is a nice starting point.  The
+// cache width must be at least 1.
+func NewCachingFactory(cacheSize, cacheWidth int) Factory {
+	if cacheSize < 1 {
+		panic("cacheSize must be at least 1")
+	}
+	if cacheWidth < 1 {
+		panic("cacheWidth must be at least 1")
+	}
```
Why not return an error instead? I think it warrants a comment

djmitche replied:
These are programming errors, rather than runtime errors -- in other words, these are invariants.  I added them only because (at least in one iteration of this code) they would otherwise cause an unterminated loop.  I could remove them, but I don't think this method should return an error if that error can practically never occur.

mx-psi replied:
I admit I just find it scary to have panics on a function in the public API of the module, but it's true that this values will be hardcoded in the program in any case I can think of, so we can leave it like this

djmitche replied:
Yeah, I understand the feeling!  It'd be nice if there was a way to do this only in dev builds, like rust's `debug_assert!(..)`.

The minima are also in the docs.  I think there are cases like this in the stdlib, too -- "must not be copied after initialized", "Reset must be called before use", etc.  Some of those lead to UB (which could include a panic), and some explicitly panic.

gh123man replied:
+1. `newTagsCache` is already returning errors in a similar way so it would be consistent. Plus since this functiona already returns an `error` (that the caller has to handle) it seems pretty low impact to move these panics to errors.

mx-psi commented on pkg/tagset/null_factory_test.go:
```diff
@@ -0,0 +1,121 @@
+
+// A nullFactory caches nothing.
+//
+// This type implements Factory.
```
Maybe better to have this instead of a comment
```go
var _ Factory = (*nullFactory)(nil)
```

I think it leads to cleaner errors and it also serves as documentation. (Maybe even a good idea to have this on every `Factory` implementation actually)

djmitche replied:
Oh, good point!

mx-psi commented on pkg/tagset/factory.go:
```diff
@@ -0,0 +1,97 @@
+
+package tagset
+
+// The Factory type is responsible for creating new Tags instances. Its
```
This interface can't be implemented outside of the module, since it has unexported methods. I think we should state that explicitly in the interface docs.

mx-psi commented on pkg/tagset/threadsafe_factory.go:
```diff
@@ -0,0 +1,102 @@
+func (f *threadsafeFactory) NewTags(src []string) *Tags {
+	f.Lock()
+	tags := f.Factory.NewTags(src)
+	f.Unlock()
```
Is there a reason to use this pattern instead of a `defer`? I know we prefer this to `defer f.Unlock()` in some situations but I am not sure why.

djmitche replied:
I think `defer` still has a runtime cost in Go1.16 -- though I'm not sure where that's documented.  In this case, the usage is formulaic enough, and the control flow completely linear, that it's easy to avoid.  I've preferred defer in other cases.

djmitche replied:
Correction: this is the _only_ use of sync.Mutex in this package.  I've preferred defer elsewhere in other PRs, though!

djmitche replied:
https://github.com/golang/go/issues/14939

remeh replied:
A note that on such methods, with recent version of Go you could consider you fine. But I agree that `defer` should not always be use just because it exists, but we should use it when it's needed and when it is clarifying the reading, which is not the case here (but my personal opinion is bleeding in here).

mx-psi reviewed (commented):
I haven't checked everything yet, but I had some comments on a first round

djmitche committed:
(fixup) use 'oldtagset' to refer to pkg/tagset/old

djmitche committed:
add interface assertions

djmitche committed:
include RFC in the repo, update docs

djmitche committed:
add comments about builder caches

urseberry reviewed (commented):
No docs review required.

djmitche committed:
remove TODO notes about efficiency

remeh commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,179 @@
+	})
+}
+
+// DisjointUnion implements Factory.DisjointUnion
```
OK I've just seen in the `Factory` comments that you mention this.

I'm wondering if there is any use-case where the caller will know for sure that there is no duplicates.

This method is slightly dangerous, we should maybe emphase (in the factory comment?) that it is the caller responsibility. I mean, you already did it, but maybe emphase it more, because it could lead to undefined behavior and if the mistake is made by someone, it will very hard to troubleshoot since there is no error message nor warning nor anything (obv. for perf reasons).

djmitche replied:
As noted in another comment, I renamed this to UnsafeDisjointUnion.  The "Unsafe" will serve as a marker that use of this method deserves careful consideration and review.

remeh commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,192 @@
+	// reallocate the storage if there is not room in the existing array, and
+	// copy the existing data into the new array
+	if cap(bldr.tags) < newLen {
+		newTags = make([]string, newLen, newLen*2)
+		newHashes = make([]uint64, newLen, newLen*2)
```
Will that be more efficient than relying on `append` on slices?

djmitche replied:
We don't want to append in this case, since the order of the elements is important in the SliceBuilder.  This code resizes and then inserts.

The Builder does not care about order, and just uses append.

remeh replied:
Got it.

remeh commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,192 @@
+	var newTags []string
+	var newHashes []uint64
+
+	// reallocate the storage if there is not room in the existing array, and
```
typo `not`

remeh commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,192 @@
+// The Add methods cannot be called after freezing.
+func (bldr *SliceBuilder) FreezeSlice(a, b int) *Tags {
+	// free unnecessary memory, and as a convenient side-effect ensure a panic
+	// if any of the Add methods are used after this point
```
It introduces the possibilities of "runtime" panic (e.g. after 2h of a running Agent) which I would not call convenient.

However I get the idea that once `Freeze`/`FreezeSlice` has been called, `Add*` should error, but, a panic in this situation is dangerous IMO.

djmitche replied:
(discussed via voice) This is an example of a runtime panic from a programming error.  Ideally this panic would happen every time, such that any amount of test coverage of the caller's code would trigger the panic.  Similar things occur in the stdlib, such as `net/http`:
```
	// read fields of the request in a separate goroutine. Callers
	// should not mutate or reuse the request until the Response's
	// Body has been closed.
```

Requirements like this rely on the fallible programmer (and reviewer), and are often the source of bugs, but it's the tool we have.  What I can do to make this easier is exhort the caller to use builders start-to-finish in a single method, and provide an example of that.

djmitche replied:
I don't think this is an issue, so there's no need to add more API surface for it.  I can remove the comment about the panic, and just include the part about saving memory.

remeh commented on pkg/tagset/old/sort_uniq.go:
```diff
@@ -0,0 +1,58 @@
+
+import "sort"
+
+// NOTE: this functionality is copied from ./pkg/util.  It will probably not be
```
Is it duplicated because of cyclic imports?

djmitche replied:
Worse -- it's because `./pkg/tagset` is a module, which does not depend on `./`, which is the module containing `./pkg/util`.

remeh commented on pkg/tagset/tags.go:
```diff
@@ -0,0 +1,154 @@
+)
+
+// Tags contains a set of tags, along with a 64-bit hash considered to be
+// unique to that set of tags.  Tags in the tagset are unique.  The order of
```
> Tags in the tagset are unique.

Nothing is enforcing it, I would not detail it as a characteristic of the struct this way.

djmitche replied:
(discussed via voice) It's true that the caller can break this invariant, but only by failing to follow the requirements in the docstrings -- in particular, but calling `DisjointUnion` with sets that are not actually disjoint.  So we'll leave this invariant in place, but I'll improve the docstring for DisjointUnion, and rename it to UnsafeDisjointUnion.

remeh commented on pkg/tagset/tags_cache.go:
```diff
@@ -0,0 +1,106 @@
+	untilRotate int
+
+	// tagests contains the constitutent tagset maps. This is a slice of length
+	// cacheCount.  The first map is the newest, into which new values will be
```
I know why you're doing this two spaces thing but I can't help my eyes to not catch it as a typo 😅

djmitche replied:
Sorry!  `sed -i -e 's/\.  /. /g' $(git grep -l '\.  ')` fixes it up..

remeh replied:
it was a silly remark, please ignore ofc.

remeh commented on pkg/tagset/union_cache_key.go:
```diff
@@ -0,0 +1,56 @@
+	var buf [16]byte
+	binary.LittleEndian.PutUint64(buf[:8], aHash)
+	binary.LittleEndian.PutUint64(buf[8:], bHash)
+	return murmur3.Sum64(buf[:])
```
Should we give a shot at millions of random entries for a couple hours to validate that the increased rate of collisions here isn't too high?

djmitche replied:
I ran a few million in a Python script, with zero collisions (along with ^ and + which did much worse).  I considered adding some fuzzing here, but I'm not sure it's a useful allocation of CPU/time -- if the fuzzing uses a fixed random seed, then it just tests the same million values over and over, while if it uses a different random seed on every invocation then some poor dev's PR is going to get a failure because they hit one of the known collisions.

remeh reviewed (commented):
Great work and addition to the codebase! However, I'm slightly scared about the amount of new public APIs, while I think only a few will have a usage in the codebase of the Agent.

Also, with the amount of automatic allocation / relocation / re-allocation (not hidden but happening behind the scenes, which is ultimately happening while offering an API), we may observe some surprises in the Agent use-cases. We will have to carefully monitor the first usages of this new API against real-word scenarios.

>   As indicated in the commit message, this new code is completely unused elsewhere in the codebase. The existing HashedTags code, which has a similar but not identical API, is moved to ./pkg/tagset/old. The next step will be to modify the consumers of that API to use pkg/tagset instead.

Would it be technically possible to completely remove this implementation to avoid confusion, switch to the new one? (it may be the plan).

Not approving just yet but please don't hesitate to ping me for a second pass as soon as needed.

djmitche committed:
single space after a full stop

djmitche committed:
Flag DisjointUnion as unsafe

djmitche committed:
clearer docs about proper use of builders

djmitche commented on pkg/tagset/builder.go:
```diff
@@ -0,0 +1,112 @@
+//     2. frozen (begins on call to bldr.Freeze).
+//     3. closed (begins on call to bldr.Close).
+//
+// The Add* methods may only be called in the "adding tags" stage. No methods may be called in the "closed" stage.
```
The builders will panic in that case :)

djmitche committed:
fix comment typos

djmitche commented:
> Would it be technically possible to completely remove this implementation to avoid confusion, switch to the new one? (it may be the plan).

Yes, this is the next part.  It's not quite a drop-in replacement, so I wanted to keep it to a separate PR.

djmitche commented:
> Also, with the amount of automatic allocation / relocation / re-allocation (not hidden but happening behind the scenes, which is ultimately happening while offering an API), we may observe some surprises in the Agent use-cases. We will have to carefully monitor the first usages of this new API against real-word scenarios.

I agree -- in particular, some of the factory methods allocate a slice of hashes, only to discard it if the tagset already exists.  Depending on how much "churn" that generates, it may make sense to optimize that to store and re-use a slice for that.  The idea with this PR is to get the API down, with later PRs addressing optimizations within the API, separately from PRs addressing usage of the API from outside.

djmitche commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,179 @@
+	return f.getCachedTags(byTagsetHashCache, hash, func() *Tags {
+		// write hashes and rewrite tags based on the map
+		hashes := make([]uint64, len(tagsMap))
+		tags = tags[:len(tagsMap)]
```
No, tagsMap does not contain any values not in tags.

djmitche commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,179 @@
+
+// NewCachingFactory creates a new caching factory.  A caching factory caches
+// Tags instances when they are seen, and uses those cached values when possible
+// to reduce CPU and memory usage.
```
I agree that it's aspirational and will depend on usage patterns.  But hopefully this allows performing fewer redundant hash operations, so maybe!

djmitche commented on pkg/tagset/builder.go:
```diff
@@ -0,0 +1,112 @@
+
+// Freeze "freezes" the builder and returns the resulting tagset.  The Add methods
+// cannot be called after freezing.
+func (bldr *Builder) Freeze() *Tags {
```
I think such a comment would need to appear in many places.  What would lead the user to assume this does not allocate, while for example `factory.NewTags` has no such assumption?

Also, note that unlike `factory.NewTags`, this doesn't _always_ allocate, if there is a cached tagset with the same hash.

mx-psi commented on pkg/tagset/docs.go:
```diff
@@ -6,20 +6,33 @@
 //
-// The HashedTags type represents an _immutable_ set of tags and associated hashes.
-// It is the primary data structure used to represent a set of tags.
+// Factories
+//
+// Factories are responsible for making new Tags instances.  Beneath a simple
+// interface, factories support optimization and deduplication.  A global factory
+// is available for general use, and purpose-specific factories can be created
+// for more intensive tag operations.
+//
+// In general, factories are not thread-safe
+//
+// Builders
+//
+// Builders are used to build tagsets tag-by-tag, before "freezing" into one or
+// more Tags instances.
```
nit: Reading this with `godoc` it looks like the 'In general, factories are not thread-safe' is shown as a title.

![image](https://user-images.githubusercontent.com/5502710/144638519-80baaa74-93d6-41f0-9b4a-5142101cd13c.png)

djmitche replied:
Ugh, I know the Go authors hate Markdown, but this is so much worse.

djmitche replied:
Also, thanks for checking the docs!!!

djmitche committed:
add fuzzing for union_cache_key

djmitche committed:
add full-stop in doc sentence

mx-psi reviewed (approved):
Thanks for the detailed docs and examples, public API looks good!

vickenty commented on docs/proposals/agent-pkg-tagset-rfc.md:
```diff
@@ -0,0 +1,303 @@
+The SliceBuilder type supports creation of overlapping "slices" of tags.
+This supports Tagger's tag-cardinality functionality while saving some memory space.
+For example, a SliceBuilder can produce a Tags containing high, orchestrator, and low-cardinality tags as well as a Tags with only the low-cardinality tags, sharing that space.
+Adding the same tag twice to a SliceBuilder is a no-op if both are at the same level, but results in undefined behavior if the levels differ.
```
Let's not have undefined behavior :) For one, UB is a scary phrase if taken in the C sense, and even if not, it isn't very helpful.

I am also curious about this level concept. What does it do? It doesn't seem to be defined in the RFC as far as I can tell. How will it be used?

djmitche replied:
SliceBuilder is described [here in the RFC](https://github.com/DataDog/architecture/blob/master/rfcs/agent-pkg-tagset/rfc.md#builders).

This is used by the tagger, which is currently not very perf-sensitive, so I can add some checks for adding the same tag twice (it just means keeping some extra maps and checking/inserting into them).

gh123man replied:
> I am also curious about this level concept. What does it do? It doesn't seem to be defined in the RFC as far as I can tell. How will it be used?

+1 I am still reading the code but I am not finding a clear definition of what a `level` and how it's intended to be used.

djmitche replied:
Can you indicate what's missing from my previous comment?

djmitche replied:
Here's an example of how it's used: https://github.com/DataDog/datadog-agent/blob/23cb0c5fd9b3bc1b420ca7e810786c52926662b5/pkg/tagger/tagstore/entity_tags.go#L124-L138

vickenty commented on docs/proposals/agent-pkg-tagset-rfc.md:
```diff
@@ -0,0 +1,303 @@
+
+A builder goes through three stages in its lifecycle: 
+ 1. adding tags (begins on `factory.New*Builder`).
+ 2. frozen (begins on call to `acc.Freeze*`).
```
```suggestion
 2. frozen (begins on call to `builder.Freeze*`).
```

vickenty commented on pkg/aggregator/aggregator.go:
```diff
@@ -17,7 +17,7 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/tagger"
 	"github.com/DataDog/datadog-agent/pkg/tagger/collectors"
-	"github.com/DataDog/datadog-agent/pkg/tagset"
+	oldtagset "github.com/DataDog/datadog-agent/pkg/tagset/old"
```
Let's make new tagset impl `pkg/tagset/v2` and leave old usage intact until we are really ready to change it. WDYT?

vickenty commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// ParseDSD implements Factory.ParseDSD
+func (f *cachingFactory) ParseDSD(data []byte) (*Tags, error) {
```
How this performs compared to the current parser and interner?

vickenty commented on pkg/tagset/null_factory_test.go:
```diff
@@ -0,0 +1,121 @@
+	testFactory(t, func() Factory { return newNullFactory() })
+}
+
+// A nullFactory caches nothing. It is useful for tests that need a factory.
```
Why tests would not be able to use a caching factory instead?

djmitche replied:
This helps to isolate behavior in tests (so that cached results aren't used by mistake, for example).

sgnn7 replied:
Would it be useful to have a reset on the caching factory internal state and avoid the duplicated code here?

vickenty commented on pkg/tagset/factory.go:
```diff
@@ -0,0 +1,105 @@
+	// This function is only useful in rare circumstances, and its use should
+	// be accompanied by a comment explaining why the usage is safe.  In
+	// general, prefer to use Union.
+	UnsafeDisjointUnion(a, b *Tags) *Tags
```
Is there a circumstance where we know we will need this? Otherwise I think we shouldn't include this at all.

vickenty commented on pkg/tagset/null_factory_test.go:
```diff
@@ -0,0 +1,121 @@
+	return f.NewTagsFromMap(tags)
+}
+
+// UnsafeDisjointUnion implements Factory.DisjoingUnion
```
Does the comment need updating?

vickenty commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,179 @@
+
+	// Tags instances are cached by 64-bit cache keys that can have a range of
+	// meanings; each CacheID identifies a different meaning. Caches with
+	// different CacheIDs are stored independently.
```
Why do we need to shard them by meaning?

ogaca-dd reviewed (commented):
Great PR!
I have added few comments.

Is there any plan to benchmark this code before using it everywhere?

djmitche commented:
Capturing some sync discussion here:
 * I'll get rid of DisjointUnion, as it's probably unnecessary and causing concern
 * I'll get rid of ParseDSD and the JSON and YAML support, as they can be added when they are needed
 * I need to implement Phase 1B (using this API to replace pkg/tagset/old) and benchmark that to demonstrate that it's more efficient than what we're running right now.  At that point, I might as well land the whole thing
 * (and comments above)

djmitche commented on pkg/tagset/builder.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+// Close closes the builder, freeing its resources.
+func (bldr *Builder) Close() {
```
I did this in the interest of a parallel construction with SliceBuilder, but that's not really useful -- particularly since SliceBuilder will probably only be used in the Tagger.

djmitche commented on pkg/tagset/union_cache_key.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/twmb/murmur3"
+)
+
+// unionCacheKey calculates a cache key that can be used to memoize Union(a, b)
```
Yes, this is critical to the combination of tags from the Tagger and Dogstatsd.

djmitche replied:
For example, here: https://github.com/DataDog/datadog-agent/blob/23cb0c5fd9b3bc1b420ca7e810786c52926662b5/pkg/aggregator/aggregator.go#L441-L452

djmitche commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,192 @@
+// Add adds the given tag to the builder at the given level.  If the tag is
+// already in the builder, it is not added again (regardless of the existing
+// tag's level).
+func (bldr *SliceBuilder) Add(level int, tag string) {
```
This will be used in the tagger, to share storage for tags of varying cardinality.

djmitche replied:
https://github.com/DataDog/datadog-agent/blob/182472b1e807258428426ef12fdc7bb56a8fe400/pkg/tagger/tagstore/entity_tags.go#L78-L139

djmitche replied:
Here's how it's used: https://github.com/DataDog/datadog-agent/blob/23cb0c5fd9b3bc1b420ca7e810786c52926662b5/pkg/tagger/tagstore/entity_tags.go#L124-L138

djmitche commented on pkg/tagset/tags.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// Contains checks whether the given tag is in this tagset.
+func (tags *Tags) Contains(tag string) bool {
```
There are a few places in the code where tags are queried for information.  Maybe it's not a good idea, but it's done and this API needs to support it.

djmitche commented on pkg/tagset/tags.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// IsSubsetOf checks whether this tagset is a subset of another tagset.
+func (tags *Tags) IsSubsetOf(other *Tags) bool {
```
It's not anything performance-sensitive.  I looked for everywhere tags are used in the repo, and at least one place was checking for a subset relationship -- but I don't remember why :)

djmitche replied:
I can remove this if you'd prefer.  But I don't think having an inefficient implementation in place is a problem, if it's not used in a perf-sensitive area.

ogaca-dd replied:
The user of `Tags` may assume this function has an efficient implementation and call it in a hoot path. Maybe renaming to something like `IsSubsetOfSlow` makes it clear this function should not be used in a hoot path. What do you think? It is also OK to keep the code as it is.

djmitche replied:
We seem to be good about measuring performance of performance-critical things, so I have confidence that such a user would notice "hey, this IsSubsetOf is pretty slow!" and improve the implementation when they needed it.  Those implementing something that is not performance-critical will not have an issue.  Let's leave it as-is and see how it's used later.  There's certainly no harm in landing it as-is on the feature branch, right?

djmitche commented on pkg/tagset/tags_cache.go:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	cacheCount := len(tc.maps)
+	for i := 1; i < cacheCount; i++ {
```
It will depend on the mix of tagset lifetimes, which may vary from usage to usage.  I suspect in most cases cacheCount = 2 will be the right choice, but if there's a mix of 50% one-time tagsets and 50% long-lived tagsets, then maps[1] gives a chance to "rescue" those long-lived tagsets and pull them back into maps[0].  cacheCount = 1 would be like the current string interner, just periodically deleting the entire cache -- not very efficient.

I expect this is something we'll want to tune, at first by hand and probably later automatically.  @vickenty had some measurements of this "churn rate" for one prod cluster, that can provide a starting point.

djmitche commented on pkg/tagset/threadsafe_factory.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// NewBuilder implements Factory.NewBuilder
+func (f *threadsafeFactory) NewBuilder(capacity int) *Builder {
```
Good spot!

djmitche replied:
I addressed this here: https://github.com/DataDog/datadog-agent/blob/e657118a2e909ee38c16319f01e5486f9bbde9af/pkg/tagset/threadsafe_factory.go#L104-L114

djmitche commented on pkg/tagset/base_factory.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+// newSliceBuilder implements NewSliceBuilder for a factory
+func (f *baseFactory) newSliceBuilder(ff Factory, levels, capacity int) *SliceBuilder {
```
The Tagger.  I think every reviewer has asked this :)

djmitche replied:
Here's the new usage: https://github.com/DataDog/datadog-agent/blob/23cb0c5fd9b3bc1b420ca7e810786c52926662b5/pkg/tagger/tagstore/entity_tags.go#L124-L138

djmitche commented on pkg/tagset/builder.go:
```diff
@@ -0,0 +1,112 @@
+	frozen  *Tags
+}
+
+// newBuilder creates a new builder.  This must be reset()
```
It's an internal API, and was simpler to implement this way -- the caller either creates or gets an existing builder, and then resets it.

djmitche commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,176 @@
+// See NewCachingFactory for usage.
+//
+type cachingFactory struct {
+	baseFactory
```
It provides some methods on the embedding struct.

djmitche commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,176 @@
+// Union implements Factory.Union
+func (f *cachingFactory) Union(a, b *Tags) *Tags {
+	key := unionCacheKey(a.Hash(), b.Hash())
+	return f.getCachedTags(byUnionHashCache, key, func() *Tags {
```
Combining tags from Dogstatsd and the Tagger.

djmitche replied:
For example, here https://github.com/DataDog/datadog-agent/blob/23cb0c5fd9b3bc1b420ca7e810786c52926662b5/pkg/aggregator/aggregator.go#L441-L452

djmitche commented on pkg/tagset/unmarshal.go:
```diff
@@ -0,0 +1,85 @@
+
+	tags, err := factory.getCachedTagsErr(byJSONCache, murmur3.Sum64(data), func() (*Tags, error) {
+		// unmarshal the underlying array of strings
+		err := json.Unmarshal(data, &bldr.buffer)
```
I was following the unmarshaling pattern here, but that's another option.  Anyway, I'm removing the marshaling/unmarshaling support.

djmitche committed:
minor fixes to succinctness

djmitche committed:
simpler Builder API: no Freeze

djmitche committed:
return an error for 0-sized caches

djmitche committed:
remove confusing disjoint union operation

djmitche committed:
remove inefficient ParseDSD method

djmitche committed:
remove JSON, YAML support

djmitche committed:
synchronize builder access in threadsafeBuilder

djmitche committed:
Merge branch 'main' into dustin.mitchell/tagset-impl

djmitche commented:
I think I've addressed all of the comments now.  I'll work on "Phase 1B", which is to use this to replace the pkg/tagset/old functionality, and benchmark the before/after.

djmitche committed:
remove comment about panic

djmitche committed:
go mod tidy

djmitche committed:
Do not re-use the slice given to NewTags

djmitche committed:
Reset slice-builder properly

djmitche committed:
recommend not using nil *tagset.Tags

djmitche committed:
Add MarshalJSON support

This is necessary for a few of the metric types where we still use JSON
payloads.

djmitche committed:
Merge branch 'main' into dustin.mitchell/tagset-impl

djmitche commented:
I found two aliasing bugs while working on #10319: cf8a1a3 and  aeeb7a3.  Avoiding aliasing is hard!!  I also noticed that MarshalJSON support will be necessary (and it's trivially simple, anyway) so I added that back in.  And I've merged from master.

djmitche committed:
Merge remote-tracking branch 'upstream/main' into dustin.mitchell/tagset-impl

djmitche commented:
OK, I've changed the base for this PR to be the topic branch in https://github.com/DataDog/datadog-agent/pull/10363.  I'm hoping that allays some of the open questions regarding "how will this be used" and "what are the performance implications" enough that I can land it and go on to make PRs containing the answers to those questions.  @vickenty, @remeh, @gh123man, and @ogaca-dd please review again!

djmitche committed:
Remove implementation process from tagset proposal

We will not be following this process.

djmitche committed:
add more how-is-this-used information to proposal

djmitche committed:
remove functionality from the proposal that is no longer included in the implementation

djmitche commented:
I've updated the proposal to better describe the current implementation.  It represents a good summary of this functionality.

djmitche committed:
Revert making pkg/tagset a module

This is out of concern that making the package a module might lead to
compatibility issues later.  This change can be reverted when required.

djmitche committed:
Merge branch 'main' into dustin.mitchell/tagset-impl

djmitche commented on pkg/tagset/tags_cache.go:
```diff
@@ -0,0 +1,106 @@
+	maps []map[uint64]*Tags
+}
+
+func newTagsCache(insertsPerRotation, cacheCount int) tagsCache {
```
I added this.

djmitche commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,176 @@
+func (f *cachingFactory) Union(a, b *Tags) *Tags {
+	key := unionCacheKey(a.Hash(), b.Hash())
+	return f.getCachedTags(byUnionHashCache, key, func() *Tags {
+		tags := make(map[string]struct{}, len(a.tags)+len(b.tags))
```
That's fine -- we can optimize in follow-up PRs.

djmitche committed:
return error instead of panicking

djmitche commented:
Hm, GitHub hasn't done a good job of filtering out the merge commit :(

I'm not sure how best to proceed.  I can
 * tough it out
 * rebase onto main
 * rebase onto main and open a new PR
 * something else...
 
 WDYT?

gh123man reviewed (approved):
Looks great! Excited to see more of how this is used and what the perf is like.

vickenty commented:
> Hm, GitHub hasn't done a good job of filtering out the merge commit :(

I think this is because your target branch does not have the same commits from main. Updating it might help.

djmitche commented:
Indeed, I forgot I had changed the base of this PR.  Fixed now,

ogaca-dd commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// Close closes the builder, freeing its resources.
+func (bldr *SliceBuilder) Close() {
```
Can `Freeze` and `Close` be merged?

djmitche replied:
In `Builder`, they are.  But not in this case (`SliceBuilder`), since multiple slices can be frozen before the builder is closed.

ogaca-dd commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,186 @@
+
+	// Offsets into the tags/hashes arrays, such that level L is represented as
+	// tags[offsets[L]:offsets[L+1]]. This slice has length (levels+1).
+	offsets []int
```
Instead of having `offsets` and a single `tags []string` why not having `tagsByLevel [][]string` (First dimension is the level index). I think it would make the implementation easier to understand with similar performance.

djmitche replied:
The FreezeSlice operation may return tags from multiple levels, so if we collected tags in a two-level slice as you suggest, that data structure would later need to be "flattened" to a one-level slice to support FreezeSlice.  That'd be several additional allocations (one per level) and involve copying every element once, so I don't think it would be more efficient than the existing insertion-sort-based approach.

ogaca-dd replied:
Indeed, I did not notice there is no slice allocation in `FreezeSlice`

ogaca-dd commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,186 @@
+//
+// In general, the easiest way to ensure these stages are followed (and allow reviewers
+// to verify this) is to use a builder in a single method, as shown in the example.  Avoid
+// storing builders in structs.
```
What do you think about adding an explanation that this struct stores tags by level and allows to merge them using a level range?

djmitche replied:
Ah, good point, thanks!

djmitche committed:
add better docs for SliceBuilder

sgnn7 commented on pkg/tagset/default.go:
```diff
@@ -0,0 +1,46 @@
+var EmptyTags *Tags
+
+func init() {
+	cf, _ := NewCachingFactory(1000, 5)
```
Should 1000 be a tunable config value?

djmitche replied:
Per discussion today, I replaced this with a non-caching factory.  Eventually, yes, it may be (or perhaps will tune itself).

sgnn7 commented on pkg/tagset/default_test.go:
```diff
@@ -0,0 +1,45 @@
+package tagset
```
Missing copyright

sgnn7 commented on pkg/tagset/default_test.go:
```diff
@@ -0,0 +1,45 @@
+		NewTags([]string{"c", "d", "e"}),
+	)
+	tags.validate(t)
+}
```
Some tests around the thread-safety claim of the class under test may be good to add here

djmitche replied:
Done.

sgnn7 commented on pkg/tagset/factory.go:
```diff
@@ -0,0 +1,85 @@
+	byTagsetHashCache cacheID = iota
+	// byUnionCache indexes Union(a,b) by uint64(a.Hash() + b.Hash())
+	byUnionHashCache cacheID = iota
+	numCacheIDs
```
```suggestion
        // numCacheIDs is the count of cache IDs
	numCacheIDs = iota
```

sgnn7 commented on pkg/tagset/fuzz_test.go:
```diff
@@ -0,0 +1,27 @@
+package tagset
```
Missing copyright

sgnn7 commented on pkg/tagset/old/sort_uniq.go:
```diff
@@ -0,0 +1,58 @@
+package tagset
```
Missing copyright

sgnn7 commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,197 @@
+// each tag, and allows the creation of Tags instances that include tags with
+// specific levels, and those Tags instances can share storage.  For example,
+// this is useful when generating tags of different cardinalities (mapping
+// cardinalities to levels).  Then the low-cardinality tags can be sliced off
```
```suggestion
// cardinalities to levels). Then the low-cardinality tags can be sliced off
```

sgnn7 commented on pkg/tagset/slice_builder.go:
```diff
@@ -0,0 +1,197 @@
+		copy(newTags[insertAt+1:], bldr.tags[insertAt:])
+
+		copy(newHashes[:insertAt], bldr.hashes[:insertAt])
+		copy(newHashes[insertAt+1:], bldr.hashes[insertAt:])
```
While I fully acknowledge that this `Add` implementation is highly optimized and very clean (subjectively speaking from my pov), I'm doubtful that this code will be very forward-maintainable due to the high amount of low-level work with offset/item shifting, flat tag array co-mingled levels, and exponential resizing all in the same file. I would personally advocate at least for careful decomposition here into logical functions even if the price of additional function calls will reduce performance to ensure that we can easily maintain it in the future.

djmitche replied:
Done!

sgnn7 commented on pkg/tagset/slice_builder_test.go:
```diff
@@ -0,0 +1,119 @@
+package tagset
```
Missing copyright

sgnn7 commented on pkg/tagset/tags.go:
```diff
@@ -0,0 +1,145 @@
+
+// Sorted returns a copy of the tags in this tagset, sorted. This is intended
+// for ease of assertions in tests, and not for use in production code.
+func (tags *Tags) Sorted() []string {
```
If we use it just in tests, could we make it private?

djmitche replied:
It's used in tests all over the codebase, so that's not possible.

sgnn7 commented on pkg/tagset/tags_cache.go:
```diff
@@ -0,0 +1,114 @@
+package tagset
```
Missing copyright

sgnn7 commented on pkg/tagset/tags_cache.go:
```diff
@@ -0,0 +1,114 @@
+	// number of inserts remaining before next rotation
+	untilRotate int
+
+	// tagests contains the constitutent tagset maps. This is a slice of length
```
```suggestion
	// maps contains the constituent tagset maps. This is a slice of length
```

sgnn7 commented on pkg/tagset/tags_cache_test.go:
```diff
@@ -0,0 +1,178 @@
+package tagset
```
Missing copyright

sgnn7 commented on pkg/tagset/threadsafe_factory_test.go:
```diff
@@ -0,0 +1,21 @@
+		cf, _ := NewCachingFactory(10, 5)
+		return NewThreadsafeFactory(cf)
+	})
+}
```
Do we need thread-safe specific tests here too?

djmitche replied:
Done.

sgnn7 commented on pkg/tagset/union_cache_key.go:
```diff
@@ -0,0 +1,56 @@
+package tagset
```
Missing copyright

sgnn7 reviewed (commented):
@djmitche Good improvements all around! I know I'm late on the review of this PR and others have reviewed it already so feel free to skip things if needed.

djmitche committed:
Do no tagset caching by default.

This is to avoid surprises in memory usage when this API is used in
non-performance-critical code.

djmitche commented on pkg/tagset/base_factory.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func (f *baseFactory) builderClosed(builder *Builder) {
+	f.builders = append(f.builders, builder)
```
The reset operation potentially _allocates_.  Builders already release unused memory on Close.

I don't think we need to worry about the unbounded nature -- they are bounded by total concurrent use of builders (that is, the  maximum length of `f.builders` is the maximum number of goroutines concurrently between a `NewBuilder` call and a `builder.Close` call).

djmitche committed:
Add missing copyright headers

djmitche committed:
prefer one space after full-stop in text

djmitche committed:
hyphenate thread-safe

djmitche commented on pkg/tagset/caching_factory.go:
```diff
@@ -0,0 +1,156 @@
+
+	var caches [numCacheIDs]tagsCache
+	for i := range caches {
+		var err error
```
Discussed on Slack -- it's fine as-is.

djmitche commented on pkg/tagset/builder.go:
```diff
@@ -0,0 +1,98 @@
+	return &Builder{
+		factory: factory,
+		tags:    nil,
+	}
```
We do always reset when creating a builder:
https://github.com/DataDog/datadog-agent/blob/6b04791007ae0104d5f8068ee7fcfa5de383d3f0/pkg/tagset/base_factory.go#L17-L28
but I think you're asking why this isn't done in the constructor.  I could reorganize it so that the constructor takes a capacity and reset's to that capacity, but then we either double-reset or the NewBuilder function becomes more complex.  I don't think that solution is any more maintainable than what I've got now, but it's not worse.  I'm happy to make that change.

djmitche replied:
Done.

djmitche committed:
reorganize resetting of builders

djmitche committed:
fuzz thread-safety of threadSafeFactory

djmitche committed:
fuzz thread-safety of the default factory

djmitche committed:
break SliceBuilder#Add into functions

djmitche committed:
fix typo

djmitche committed:
use a Rand per goroutine
